openff/toolkit/topology/_mm_molecule.py

```diff
@@ -323,18 +327,55 @@ def _is_isomorphic_with(self, other) -> bool:
         This currently does NOT checks that the two molecules have
         * Topologically identical bond graphs
         """
+        return _SimpleMolecule.are_isomorphic(
+            self,
+            other,
+            return_atom_map=False,
+        )[0]
+
+    @staticmethod
+    def are_isomorphic(
+        mol1: Union["FrozenMolecule", "_SimpleMolecule", "nx.Graph"],
+        mol2: Union["FrozenMolecule", "_SimpleMolecule", "nx.Graph"],
+        return_atom_map: bool = False,
+    ) -> Tuple[bool, Optional[Dict[int, int]]]:
+        import networkx as nx
+
+        _cls = _SimpleMolecule
 
-        if self.n_atoms != other.n_atoms:
-            return False
+        if isinstance(mol1, nx.Graph) and isinstance(mol2, nx.Graph):
+            mol1 = _SimpleMolecule._from_subgraph(mol1)
+            mol2 = _SimpleMolecule._from_subgraph(mol2)
 
-        if self.n_bonds != other.n_bonds:
-            return False
+        elif isinstance(mol1, nx.Graph):
+            assert isinstance(mol2, _cls)
+            mol1 = _SimpleMolecule._from_subgraph(mol1)
 
-        for this_atom, other_atom in zip(self.atoms, other.atoms):
+        elif isinstance(mol2, nx.Graph):
+            assert isinstance(mol1, _cls)
+            mol2 = _SimpleMolecule._from_subgraph(mol2)
+
+        else:
+            # static methods (by definition) know nothing about their class,
+            # so the class to compare to must be hard-coded here
+            if not (isinstance(mol1, _cls) and isinstance(mol2, _cls)):
+                return False, None
+
+        if mol1.n_atoms != mol2.n_atoms:
+            return False, None
+
+        if mol1.n_bonds != mol2.n_bonds:
+            return False, None
+
+        for this_atom, other_atom in zip(mol1.atoms, mol2.atoms):
             if this_atom.atomic_number != other_atom.atomic_number:
-                return False
+                return False, None
 
```

Comment:
(blocking) For the use cases I can foresee, false negatives in this check are fairly benign (just lead to duplicated atom types), but false positives would be really bad (could silently lead to incorrect energies). I'm pretty certain this could return a false positive when comparing, for example, n-propanol and iso-propano, if both happen to be indexed starting with CCCO. 

So this should probably also evaluate the bond graph by comparing something like `sorted([neighbor.molecule_atom_index for neighbor in this_atom.bonded_atoms()])` for each atom.

I hate to call this blocking because the previous implementation also had this issue, but leaving this in its current state will very likely lead to a Class I Silent Wrong Energy Bug sooner or later.

Reply:
I've implemented a different check - atomic number of each bonded neighbor of each atom, accounting for order - that passes the propanol case. I'm not sure in which other cases two of these types of molecules might no be isomorphic but pass these requirements

Reply:
I think there's still a plausible edge case where this could return a false positive. I tried to type it up but realized it would be quicker to just commit a minimal reproducing test. The edge case is where two mols could seem isomorphic via the previous logic depending on the path used to set the indexing. Since SMILES are impossible to read, here's a picture of what the new test compares (with the C atoms labeled with their indices)

<img width="974" alt="Screen Shot 2023-04-21 at 10 29 21 AM" src="https://user-images.githubusercontent.com/16329175/233699710-fe3b9e3e-5a28-40f1-8f01-00de63b2c592.png">